materialize-mysql/driver.go

```diff
@@ -0,0 +1,620 @@
+			if details, ok := varchars[col.Identifier]; ok {
+				columnMetas[idx] = varcharColumnMeta{
+					identifier: col.Identifier,
+					maxLength:  details.maxLength,
+				}
+
+				log.WithFields(log.Fields{
+					"table":            b.target.Identifier,
+					"column":           col.Identifier,
+					"varcharMaxLength": details.maxLength,
+					"collection":       b.target.Source.String(),
+					"field":            col.Field,
+				}).Info("matched string collection field to table VARCHAR column")
+			}
+		}
+	}
+	b.varcharColumnMetas = columnMetas
+
+	t.bindings = append(t.bindings, b)
+
+	// Create a binding-scoped temporary table for staged keys to load.
+	if _, err := t.load.conn.ExecContext(ctx, b.createLoadTableSQL); err != nil {
+		return fmt.Errorf("Exec(%s): %w", b.createLoadTableSQL, err)
+	}
+
+	return nil
+}
+
+func (d *transactor) Load(it *pm.LoadIterator, loaded func(int, json.RawMessage) error) error {
+	var ctx = it.Context()
+
+	// Use a read-only "load" transaction, which will automatically
+	// truncate the temporary key staging tables on commit.
+  var txn, err = d.load.conn.BeginTx(ctx, &stdsql.TxOptions{ReadOnly: true})
+	if err != nil {
+		return fmt.Errorf("DB.BeginTx: %w", err)
+	}
+	defer txn.Rollback()
+
+	for it.Next() {
+		var b = d.bindings[it.Binding]
+
+		if converted, err := b.target.ConvertKey(it.Key); err != nil {
+			return fmt.Errorf("converting Load key: %w", err)
+		} else {
+			txn.ExecContext(ctx, b.loadInsertSQL, converted...)
+		}
+	}
+
+	if it.Err() != nil {
+		return it.Err()
+	}
+
+	// Issue a union join of the target tables and their (now staged) load keys,
+	// and send results to the |loaded| callback.
+	rows, err := txn.QueryContext(ctx, d.load.unionSQL)
+	if err != nil {
+		return fmt.Errorf("querying Load documents: %w", err)
+	}
+	defer rows.Close()
+
+	for rows.Next() {
+		var binding int
+		var document json.RawMessage
+
+		if err = rows.Scan(&binding, &document); err != nil {
+			return fmt.Errorf("scanning Load document: %w", err)
+		} else if err = loaded(binding, json.RawMessage(document)); err != nil {
+			return err
+		}
+	}
+	if err = rows.Err(); err != nil {
+		return fmt.Errorf("querying Loads: %w", err)
+	} else if err = txn.Commit(); err != nil {
+		return fmt.Errorf("commiting Load transaction: %w", err)
+	}
+
+	return nil
+}
+
+func (d *transactor) Store(it *pm.StoreIterator) (_ pm.StartCommitFunc, err error) {
+	ctx := it.Context()
+	txn, err := d.store.conn.BeginTx(ctx, &stdsql.TxOptions{})
+	if err != nil {
+		return nil, fmt.Errorf("DB.BeginTx: %w", err)
+	}
+	defer func() {
+		if err != nil {
+			txn.Rollback()
+		}
+	}()
+
+	// varcharColumnUpdates records any VARCHAR columns that need their lengths increased. The keys
+	// are table identifiers, and the values are a list of column identifiers that need altered.
+	// Columns will only ever be to altered it to VARCHAR(MAX).
+	varcharColumnUpdates := make(map[string][]string)
```

Comment:
I think it is still needed to keep track of the maxLength of a column, it may be resized multiple times, and we want to keep track of the maxLength as it changes while we iterate over columns

Reply:
That tracking is done by `b.varcharColumnMetas`, I don't see the `varcharColumnUpdates` map that is made here used anywhere else, other than updating it on line 577. In redshift it gets passed through to the commit function so that the column updates can be done async as part of commiting the txn.

Reply:
ah sorry, I mixed up `varcharColumnMetas` and `varcharColumnUpdates`, you are right 👍🏽